// relaen/dist/core/dialect/mssql/mssqlprovider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MssqlProvider = void 0;
const baseprovider_1 = require("../../baseprovider");
const entityfactory_1 = require("../../entityfactory");
const transactionmanager_1 = require("../../transactionmanager");
/**
 * mssql provider
 * @since 0.3.0
 */
class MssqlProvider extends baseprovider_1.BaseProvider {
    /**
     * 构造器
     * @param cfg   连接配置
     */
    constructor(cfg) {
        super(cfg);
        this.dbMdl = require('mssql');
        this.options = cfg.options ? cfg.options : {
            server: cfg.host,
            port: cfg.port,
            user: cfg.username,
            password: cfg.password,
            database: cfg.database,
            connectionTimeout: cfg.connectTimeout,
            requestTimeout: cfg.requestTimeout,
            options: {
                trustServerCertificate: true,
            }
        };
        // 连接池
        if (cfg.usePool || cfg.pool) {
            if (!cfg.options && cfg.pool) {
                this.options['pool'] = {
                    max: cfg.pool.max,
                    min: cfg.pool.min,
                    idleTimeoutMillis: cfg.idleTimeout
                };
            }
            this.pool = new this.dbMdl.ConnectionPool(this.options);
            console.log(this.pool.totalCount);
        }
    }
    /**
     * 获取连接
     * @returns     数据库连接
     */
    async getConnection() {
        if (this.pool) {
            return await this.pool.connect();
        }
        return await this.dbMdl.connect(this.options);
    }
    /**
     * 关闭连接
     * @param connection    数据库连接对象
     */
    async closeConnection(connection) {
        if (!this.pool) {
            await connection.conn.close();
        }
        return null;
    }
    /**
     * 执行sql语句
     * @param connection    db connection
     * @param sql           待执行sql
     * @param params        参数数组
     * @returns             结果(集)
     */
    async exec(connection, sql, params) {
        let request; //用request作为sql执行器
        //如果事务存在，则通过事务获取request，否则通过connection获取
        let tr = transactionmanager_1.TransactionManager.get();
        if (tr) {
            request = tr.tr.request();
        }
        else {
            request = connection.conn.request();
        }
        params = params || [];
        // 绑定数组或对象参数
        if (Array.isArray(params)) {
            params.forEach((value, index) => {
                request.input(index.toString(), value);
            });
        }
        else {
            Object.getOwnPropertyNames(params).forEach((item) => {
                request.input(item, params[item]);
            });
        }
        let result = await request.query(sql);
        return result.recordset;
    }
    /**
     * 处理记录起始记录索引和记录数
     * @param sql       sql
     * @param start     开始索引
     * @param limit     记录数
     * @returns         处理后的sql
     */
    handleStartAndLimit(sql, start, limit) {
        //mssql 需要orderby，offset fetch同时使用
        handleOrder();
        if (start && limit) {
            return sql + " OFFSET " + start + " ROWS FETCH NEXT " + limit + " ROWS ONLY";
        }
        if (limit) {
            return sql + " OFFSET 0 ROWS FETCH NEXT " + limit + " ROWS ONLY";
        }
        if (start) {
            return sql + " OFFSET " + start + " ROWS";
        }
        return sql;
        //无order by 则需要添加
        function handleOrder() {
            if (!/order\s+by/i.test(sql)) {
                let r = /from\s+\w+/i.exec(sql);
                if (!r) {
                    return sql;
                }
                let tbl = r[0].replace(/from\s+/i, '');
                let t0 = /from\s+\w+\s+t0/i.test(sql);
                let orderBy = '(SELECT NULL)';
                let cfg = entityfactory_1.EntityFactory.getEntityCfgByTblName(tbl);
                if (cfg) {
                    orderBy = (t0 ? 't0.' : '') + cfg.columns.get(cfg.id.name).name + ' ASC';
                }
                sql += ' ORDER BY ' + orderBy;
            }
        }
    }
    /**
     * 获取实体sequence，针对主键生成策略为sequence时有效
     * @param em        entity manager
     * @param seqName   sequence name
     * @param schema    schema
     * @returns         sequence 值
     */
    async getSequenceValue(em, seqName, schema) {
        let query = em.createNativeQuery("select next value for " + seqName);
        let r = await query.getResult();
        if (r) {
            //转换为整数
            return parseInt(r);
        }
        return 0;
    }
    /**
     * 从sql执行结果获取identityid，仅对主键生成策略是identity的有效
     * @param result    sql执行结果
     * @returns         主键
     */
    getIdentityId(result) {
        if (!result || result.length > 1 || !result[0].insertId) {
            return;
        }
        return result[0].insertId;
    }
    /**
     * 获取加锁sql语句
     * @param type      锁类型
     * @param tables    表名，表锁时使用
     * @param schema    模式名，表锁时使用
     * @retruns         加锁sql语句
     * @since           0.4.0
    */
    lock(type, tables, schema) {
        if (schema && tables) {
            tables.forEach((v, i) => {
                tables[i] = schema + '.' + tables[i];
            });
        }
        switch (type) {
            case 'table_read':
                return "SELECT * FROM " + tables.join() + " WITH (HOLDLOCK,TABLOCK)";
            case 'table_write':
                return "SELECT * FROM " + tables.join() + " WITH (UPDLOCK,TABLOCK)";
            case 'row_read':
                return "WITH(HOLDLOCK,ROWLOCK)";
            case 'row_write':
                return "WITH(UPDLOCK,ROWLOCK)";
        }
    }
    /**
     * 获取新增返回主键字段sql语句
     * @param idField   主键字段
     * @returns         查询主键sql语句
     */
    insertReturn(idField) {
        return "SELECT @@IDENTITY AS insertId";
    }
}
exports.MssqlProvider = MssqlProvider;
