// relaen/dist/core/dialect/mssql/mssqltransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MssqlTransaction = void 0;
const connectionmanager_1 = require("../../connectionmanager");
const logger_1 = require("../../logger");
const transaction_1 = require("../../transaction");
/**
 * mssql 事务类
 * @since 0.3.0
 */
class MssqlTransaction extends transaction_1.Transaction {
    /**
     * 构造器
     * @param conn      connection
     */
    constructor(conn) {
        super(conn);
        //创建实际的transaction
        this.tr = conn.conn.transaction();
    }
    /**
     * 设置当前事务
     * @param isolation 事务隔离级
     * @since           0.4.0
     */
    async setIsolationLevel(isolationLevel) {
        if (isolationLevel) {
            let sql = "SET TRANSACTION ISOLATION LEVEL " + isolationLevel;
            await connectionmanager_1.ConnectionManager.provider.exec(this.conn, sql);
            logger_1.Logger.log(sql);
        }
    }
    /**
     * 开始事务
     */
    async begin() {
        await this.tr.begin();
        super.begin();
    }
    /**
     * 提交事务
     */
    async commit() {
        await this.tr.commit();
        super.commit();
    }
    /**
     * 事务回滚
     */
    async rollback() {
        await this.tr.rollback();
        super.rollback();
    }
}
exports.MssqlTransaction = MssqlTransaction;
