// relaen/dist/core/dialect/sqlite/sqlitetransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqliteTransaction = void 0;
const transaction_1 = require("../../transaction");
/**
 * Sqlite 事务类
 * @since 0.4.0
 */
class SqliteTransaction extends transaction_1.Transaction {
    /**
     * 设置sqlite事务类型（仅sqlite事务使用）
     * @param type  事务类型
     */
    setType(type) {
        if (type) {
            this.type = type;
        }
    }
    /**
     * 开始事务
     * @param sqliteTransaction     事务类型
     */
    async begin() {
        await new Promise((resolve, reject) => {
            let sql = this.type ? 'begin ' + this.type : 'begin';
            this.conn.conn.run(sql, (err) => {
                if (err) {
                    return reject(err);
                }
                super.begin();
                resolve(null);
            });
        });
    }
    /**
     * 事务提交
     */
    async commit() {
        await new Promise((resolve, reject) => {
            this.conn.conn.run('commit', (err) => {
                if (err) {
                    return reject(err);
                }
                super.commit();
                resolve(null);
            });
        });
    }
    /**
     * 事务回滚
     */
    async rollback() {
        await new Promise((resolve, reject) => {
            this.conn.conn.run('rollback', (err) => {
                if (err) {
                    return reject(err);
                }
                super.commit();
                resolve(null);
            });
        });
    }
}
exports.SqliteTransaction = SqliteTransaction;
