// relaen/dist/core/connection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Connection = void 0;
const connectionmanager_1 = require("./connectionmanager");
const transactionfactory_1 = require("./transactionfactory");
/**
 * 数据库连接类
 */
class Connection {
    /**
     * 构造器
     * @param conn  实际的连接
     */
    constructor(conn) {
        this.useCount = 1;
        this.setConn(conn);
    }
    /**
     * 设置实际的connection
     * @param conn      实际的连接
     * @sinace 0.3.0
     */
    setConn(conn) {
        if (!conn) {
            return;
        }
        this.conn = conn;
        this.connected = true;
    }
    /**
     * 关闭连接
     * @param force     是否强制关闭
     */
    async close(force) {
        await connectionmanager_1.ConnectionManager.closeConnection(this, force);
    }
    /**
     * 创建事务对象
     */
    createTransaction() {
        let trClass = transactionfactory_1.TransactionFactory.get();
        let con = this.conn;
        return trClass ? Reflect.construct(trClass, [this]) : null;
    }
}
exports.Connection = Connection;
