// relaen/dist/core/relaenmanager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RelaenManager = void 0;
const connectionmanager_1 = require("./connectionmanager");
const entityfactory_1 = require("./entityfactory");
const errorfactory_1 = require("./errorfactory");
const translatorfactory_1 = require("./translatorfactory");
const mssqltranslator_1 = require("./dialect/mssql/mssqltranslator");
const mysqltranslator_1 = require("./dialect/mysql/mysqltranslator");
const oracletranslator_1 = require("./dialect/oracle/oracletranslator");
const postgrestranslator_1 = require("./dialect/postgres/postgrestranslator");
const placeholderfactory_1 = require("./placeholderfactory");
const transactionfactory_1 = require("./transactionfactory");
const mysqltransaction_1 = require("./dialect/mysql/mysqltransaction");
const providerfactory_1 = require("./providerfactory");
const mssqlprovider_1 = require("./dialect/mssql/mssqlprovider");
const mysqlprovider_1 = require("./dialect/mysql/mysqlprovider");
const oracleprovider_1 = require("./dialect/oracle/oracleprovider");
const postgresprovider_1 = require("./dialect/postgres/postgresprovider");
const mssqltransaction_1 = require("./dialect/mssql/mssqltransaction");
const oracletransaction_1 = require("./dialect/oracle/oracletransaction");
const postgrestransaction_1 = require("./dialect/postgres/postgrestransaction");
const mariadbprovider_1 = require("./dialect/mariadb/mariadbprovider");
const sqliteprovider_1 = require("./dialect/sqlite/sqliteprovider");
const mariadbtransaction_1 = require("./dialect/mariadb/mariadbtransaction");
const sqlitetransaction_1 = require("./dialect/sqlite/sqlitetransaction");
const sqlitetranslator_1 = require("./dialect/sqlite/sqlitetranslator");
const mariadbtranslator_1 = require("./dialect/mariadb/mariadbtranslator");
const logger_1 = require("./logger");
/**
 * relaen 框架管理器
 */
class RelaenManager {
    /**
     * 初始化
     * @param cfg   配置文件名或配置对象
     */
    static async init(cfg) {
        //TODO 不要字符串解析文件目录了吗
        if (typeof cfg !== 'object') {
            throw errorfactory_1.ErrorFactory.getError('0001');
        }
        this.dialect = cfg.dialect || 'mysql';
        this.debug = cfg.debug || false;
        this.fileLog = cfg.fileLog || false;
        this.cache = cfg.cache === false ? false : true;
        this.fullTableOperation = cfg.fullTableOperation === true ? true : false;
        this.initProvider();
        this.initTransaction();
        this.initTranslator();
        this.initPlaceholder();
        logger_1.Logger.init(this.debug, this.fileLog);
        connectionmanager_1.ConnectionManager.init(cfg);
        //加载实体
        for (let path of cfg.entities) {
            await entityfactory_1.EntityFactory.addEntities(path);
        }
    }
    /**
     * 初始化各dialect对应的translator
     */
    static initTranslator() {
        //添加到translator工厂
        translatorfactory_1.TranslatorFactory.add('mssql', mssqltranslator_1.MssqlTranslator);
        translatorfactory_1.TranslatorFactory.add('mysql', mysqltranslator_1.MysqlTranslator);
        translatorfactory_1.TranslatorFactory.add('oracle', oracletranslator_1.OracleTranslator);
        translatorfactory_1.TranslatorFactory.add('postgres', postgrestranslator_1.PostgresTranslator);
        translatorfactory_1.TranslatorFactory.add('mariadb', mariadbtranslator_1.MariadbTranslator);
        translatorfactory_1.TranslatorFactory.add('sqlite', sqlitetranslator_1.SqliteTranslator);
    }
    /**
     * 初始化各dialect对应的provider
     */
    static initProvider() {
        providerfactory_1.ProviderFactory.add('mssql', mssqlprovider_1.MssqlProvider);
        providerfactory_1.ProviderFactory.add('mysql', mysqlprovider_1.MysqlProvider);
        providerfactory_1.ProviderFactory.add('oracle', oracleprovider_1.OracleProvider);
        providerfactory_1.ProviderFactory.add('postgres', postgresprovider_1.PostgresProvider);
        providerfactory_1.ProviderFactory.add('mariadb', mariadbprovider_1.MariadbProvider);
        providerfactory_1.ProviderFactory.add('sqlite', sqliteprovider_1.SqliteProvider);
    }
    /**
     * 初始化各dialect对应的占位符配置
     */
    static initPlaceholder() {
        //添加到placeholder工厂
        placeholderfactory_1.PlaceholderFactory.add('mssql', '@', 0);
        placeholderfactory_1.PlaceholderFactory.add('mysql', '?');
        placeholderfactory_1.PlaceholderFactory.add('oracle', ':', 0);
        placeholderfactory_1.PlaceholderFactory.add('postgres', '$', 1);
        placeholderfactory_1.PlaceholderFactory.add('mariadb', '?');
        placeholderfactory_1.PlaceholderFactory.add('sqlite', '?');
    }
    /**
     * 初始化各dialect对应的transaction
     */
    static initTransaction() {
        //添加到transaction工厂
        transactionfactory_1.TransactionFactory.add('mssql', mssqltransaction_1.MssqlTransaction);
        transactionfactory_1.TransactionFactory.add('mysql', mysqltransaction_1.MysqlTransaction);
        transactionfactory_1.TransactionFactory.add('oracle', oracletransaction_1.OracleTransaction);
        transactionfactory_1.TransactionFactory.add('postgres', postgrestransaction_1.PostgresTransaction);
        transactionfactory_1.TransactionFactory.add('mariadb', mariadbtransaction_1.MariadbTransaction);
        transactionfactory_1.TransactionFactory.add('sqlite', sqlitetransaction_1.SqliteTransaction);
    }
}
exports.RelaenManager = RelaenManager;
