// relaen/dist/core/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const logger_1 = require("./logger");
const threadlocal_1 = require("./threadlocal");
const transactionmanager_1 = require("./transactionmanager");
/**
 * 事务基类
 */
class Transaction {
    /**
     * 构造器
     * @param conn  真实连接对象
     */
    constructor(conn) {
        this.conn = conn;
        //每个transaction都新建一个threadId
        this.threadId = threadlocal_1.RelaenThreadLocal.newThreadId();
        //加入事务管理器
        transactionmanager_1.TransactionManager.add(this);
    }
    /**
     * 设置当前事务
     * @param isolation 事务隔离级
     */
    async setIsolationLevel(isolationLevel) {
    }
    /**
     * 事务开始
     * @param sqliteTransaction sqlite事务类型
     */
    async begin() {
        logger_1.Logger.log('Transaction is begun.');
    }
    /**
     * 事务提交,继承类需要重载
     */
    async commit() {
        //从事务管理器移除
        transactionmanager_1.TransactionManager.remove(this.threadId);
        logger_1.Logger.log('Transaction is committed.');
    }
    /**
     * 事务回滚,继承类需要重载
     */
    async rollback() {
        //从事务管理器移除
        transactionmanager_1.TransactionManager.remove(this.threadId);
        logger_1.Logger.log('Transaction is rolled back.');
    }
}
exports.Transaction = Transaction;
