// relaen/dist/core/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ESqliteTransactionType = exports.EIsolationLevel = exports.ELockMode = exports.ELockType = exports.EQueryType = exports.EEntityState = exports.ERelationType = void 0;
/**
 * 关系类型
 */
var ERelationType;
(function (ERelationType) {
    /**
     * 一对一关系
     */
    ERelationType[ERelationType["OneToOne"] = 1] = "OneToOne";
    /**
     * 一对多关系
     */
    ERelationType[ERelationType["OneToMany"] = 2] = "OneToMany";
    /**
     * 多对一关系
     */
    ERelationType[ERelationType["ManyToOne"] = 3] = "ManyToOne";
    /**
     * 多对多关系
     */
    ERelationType[ERelationType["ManyToMany"] = 4] = "ManyToMany";
})(ERelationType || (ERelationType = {}));
exports.ERelationType = ERelationType;
/**
 * 实体状态
 */
var EEntityState;
(function (EEntityState) {
    /**
     * 新建状态
     */
    EEntityState[EEntityState["NEW"] = 1] = "NEW";
    /**
     * 持久化状态
     */
    EEntityState[EEntityState["PERSIST"] = 2] = "PERSIST";
})(EEntityState || (EEntityState = {}));
exports.EEntityState = EEntityState;
/**
 * query类型
 */
var EQueryType;
(function (EQueryType) {
    /**
     * select
     */
    EQueryType[EQueryType["SELECT"] = 0] = "SELECT";
    /**
     * insert
     */
    EQueryType[EQueryType["INSERT"] = 1] = "INSERT";
    /**
     * update
     */
    EQueryType[EQueryType["UPDATE"] = 2] = "UPDATE";
    /**
     * delete
     */
    EQueryType[EQueryType["DELETE"] = 3] = "DELETE";
})(EQueryType || (EQueryType = {}));
exports.EQueryType = EQueryType;
/**
 * 锁机制类型
 */
var ELockType;
(function (ELockType) {
    ELockType["TABLEREAD"] = "table_read";
    ELockType["TABLEWRITE"] = "table_write";
    ELockType["ROWREAD"] = "row_read";
    ELockType["ROWWRITE"] = "row_write";
})(ELockType || (ELockType = {}));
exports.ELockType = ELockType;
;
/**
 * 锁类型
 */
var ELockMode;
(function (ELockMode) {
    ELockMode["OPTIMISTIC"] = "optimistic";
    ELockMode["PESSIMISTIC_READ"] = "pessimistic_read";
    ELockMode["PESSIMISTIC_WRITE"] = "pessimistic";
})(ELockMode || (ELockMode = {}));
exports.ELockMode = ELockMode;
/**
 * 事务隔离级别
 */
var EIsolationLevel;
(function (EIsolationLevel) {
    EIsolationLevel["SERIALIZABLE"] = "SERIALIZABLE";
    EIsolationLevel["READUNCOMMITTED"] = "READ UNCOMMITTED";
    EIsolationLevel["READCOMMITTED"] = "READ COMMITTED";
    EIsolationLevel["REPEATABLEREAD"] = "REPEATABLE READ";
})(EIsolationLevel || (EIsolationLevel = {}));
exports.EIsolationLevel = EIsolationLevel;
;
/**
 * Sqlite事务类型
 */
var ESqliteTransactionType;
(function (ESqliteTransactionType) {
    ESqliteTransactionType["DEFERRED"] = "deferred";
    ESqliteTransactionType["IMMEDIATE"] = "immediate";
    ESqliteTransactionType["EXCLUSIVE"] = "exclusive";
})(ESqliteTransactionType || (ESqliteTransactionType = {}));
exports.ESqliteTransactionType = ESqliteTransactionType;
;
