// relaen/dist/core/dialect/oracle/oracleprovider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OracleProvider = void 0;
const baseprovider_1 = require("../../baseprovider");
/**
 * oracle provider
 * @since 0.3.0
 */
class OracleProvider extends baseprovider_1.BaseProvider {
    /**
     * 构造器
     * @param cfg   连接配置
     */
    constructor(cfg) {
        super(cfg);
        this.dbMdl = require('oracledb');
        this.dbMdl.autoCommit = true;
        this.options = cfg.options ? cfg.options : {
            user: cfg.username,
            password: cfg.password,
            connectString: cfg.host + ":" + cfg.port + "/" + cfg.database,
            privilege: cfg.privilege,
            stmtCacheSize: cfg.stmtCacheSize,
            tag: cfg.tag
        };
        // 连接池
        if (!cfg.options && cfg.pool) {
            this.options['poolIncrement'] = cfg.poolIncrement;
            this.options['poolMax'] = cfg.pool.max;
            this.options['poolMin'] = cfg.pool.min;
            this.options['poolTimeout'] = cfg.idleTimeout;
        }
        if (cfg.usePool || cfg.pool) {
            this.isPool = true;
        }
    }
    /**
     * 获取连接
     * @returns     connection
     */
    async getConnection() {
        if (this.isPool) {
            if (!this.pool) {
                this.pool = await this.dbMdl.createPool(this.options);
            }
            return await this.pool.getConnection();
        }
        return await this.dbMdl.getConnection(this.options);
    }
    /**
     * 关闭连接
     * @param connection    数据库连接对象
     */
    async closeConnection(connection) {
        if (this.pool) {
            await connection.conn.release();
        }
        else {
            await connection.conn.close();
        }
        return null;
    }
    /**
     * 执行sql语句
     * @param connection    db connection
     * @param sql           待执行sql
     * @param params        参数数组
     * @returns             结果(集)
     */
    async exec(connection, sql, params) {
        // 默认自动提交
        let autoCommit = connection.autoCommit === false ? false : true;
        params = params || [];
        let r = await connection.conn.execute(sql, params, { autoCommit: autoCommit, outFormat: 4002 });
        if (r.rows) {
            return r.rows;
        }
        return r;
    }
    /**
     * 处理记录起始记录索引和记录数
     * @param sql       sql
     * @param start     开始索引
     * @param limit     记录数
     * @returns         处理后的sql
     * @since           0.2.0
     */
    handleStartAndLimit(sql, start, limit) {
        if (limit && start) {
            return sql + " OFFSET " + start + " ROWS FETCH NEXT " + limit + " ROWS ONLY";
        }
        if (limit) {
            return sql + " FETCH NEXT " + limit + " ROWS ONLY";
        }
        if (start) {
            return sql + " OFFSET " + start + " ROWS";
        }
        return sql;
    }
    /**
     * 获取实体sequence，针对主键生成策略为sequence时有效
     * @param em        entity manager
     * @param seqName   sequence name
     * @param schema    schema
     * @returns         sequence 值
     */
    async getSequenceValue(em, seqName, schema) {
        // 需要指定sequence所属schema
        let query = em.createNativeQuery("select " + (schema ? schema + "." + seqName : seqName) + ".nextval from dual");
        let r = await query.getResultList(-1, -1);
        if (r[0].NEXTVAL) {
            //转换为整数
            return parseInt(r[0].NEXTVAL);
        }
        return 0;
    }
    /**
     * 获取加锁sql语句
     * @param type      锁类型
     * @param tables    表名，表锁时使用
     * @param schema    模式名，表锁时使用
     * @retruns         加锁sql语句
     * @since           0.4.0
     */
    lock(type, tables, schema) {
        if (schema) {
            tables.forEach((v, i) => {
                tables[i] = schema + '.' + tables[i];
            });
        }
        switch (type) {
            //表连接为 ',' 逗号
            case 'table_read':
                return "LOCK TABLE " + tables.join() + " IN SHARE MODE";
            case 'table_write':
                return "LOCK TABLE " + tables.join() + " IN EXCLUSIVE MODE";
            case 'table_read':
                return "FOR UPDATE";
            case 'row_write':
                return "FOR UPDATE";
            default:
                return '';
        }
    }
}
exports.OracleProvider = OracleProvider;
